### PR TITLE
Wrap SSH handshake in a timeout

### LINKED ISSUE #2333: Timeout waiting for SSH
Packer log: https://gist.github.com/rjosephwright/d0257fba61ae7795b37d
Config: https://gist.github.com/rjosephwright/16722edd70ac4ee313da

I am unable to create any images using the amazon-ebs builder due to an error "Timeout waiting for SSH". I am running 0.8.0 on Ubuntu 15.04, although I have tried with 0.7.5 (previously working for me) and get the same failure.

During the time that packer's log reports "handshaking with SSH", I can manually log into the very same instance from the command line without any issues, using both the IP and the DNS name. It is not an issue with `ssh_username`. I have tried using `ssh_keypair_name` and `ssh_private_key_file` options with known good values and still get the same result.

### PR DESCRIPTION
**Pull request by cbednarski**
Fixes #2333

I think this may still need some resource cleanup (connection close and such) but I want to make sure the approach is right.

### CODE REVIEW

File: communicator/ssh/communicator.go [hunk 1]
```diff
@@ -273,9 +281,45 @@ func (c *comm) reconnect() (err error) {
...
+	// Default timeout to 1 minute if it wasn't specified (zero value). For
+	// when you need to handshake from low orbit.
+	var duration time.Duration
+	if c.config.HandshakeTimeout == 0 {
+		duration = 1 * time.Minute
+	} else {
+		duration = c.config.HandshakeTimeout
+	}
+
+	timeoutExceeded := time.After(duration)
```
Comment (mitchellh) [inline communicator/ssh/communicator.go]: Style nitpick: no reason to make this a variable if its not reused. It is more idiomatic to do this:

`case <-time.After(duration):`

File: communicator/ssh/communicator.go [hunk 2]
```diff
@@ -273,9 +281,45 @@ func (c *comm) reconnect() (err error) {
...
+	connectionEstablished := make(chan bool, 1)
```
Comment (mitchellh) [inline communicator/ssh/communicator.go]: If you don't use a value for a channel, and you're only using it to signal, it is more idiomatic to use a `chan struct{}`. It takes up less memory space, but also just signals that the value is useless.

File: communicator/ssh/communicator.go [hunk 3]
```diff
@@ -273,9 +281,45 @@ func (c *comm) reconnect() (err error) {
...
+
+	var sshConn ssh.Conn
+	var sshChan <-chan ssh.NewChannel
+	var req <-chan *ssh.Request
+
+	go func() {
+		sshConn, sshChan, req, err = ssh.NewClientConn(c.conn, c.address, c.config.SSHConfig)
+		connectionEstablished <- true
```
Comment (mitchellh) [inline communicator/ssh/communicator.go]: More idiomatic here to `close(connectionEstablished)`. Closed channels return immediately so it'll signal the select. It also shows in code that the channel is never used again, and that the value doesn't matter.

**Comment by mitchellh** [comment]
Nitpick on some Go stuff, but the logic looks great and so does the test :+1:

File: communicator/ssh/communicator.go (hunk 3 above)
Comment (cbednarski) [inline communicator/ssh/communicator.go]: Good tip, thanks!

**Pull request merged**